lib/pf/radius.pm

```diff
@@ -828,96 +830,275 @@ sub switch_access {
+    my $merged = { %$options, %$args };
+    $merged->{'rule_class'} = $Rules::AUTH;
+    $merged->{'context'} = $pf::constants::realm::RADIUS_CONTEXT;
+    my $attributes;
+    my $matched = pf::authentication::match2([@$sources], $merged, $$extra, \$attributes);
+
+    my $value = $matched->{values}{$Actions::TRIGGER_RADIUS_MFA} if $matched;
+    if ($value) {
+        my $mfa = pf::factory::mfa->new($value);
+        my $chi = pf::CHI->new(namespace => 'mfa');
+        if ($mfa->radius_mfa_method eq 'strip-otp') {
+            # Previously did a authentication request ?
+            if (my $infos = $chi->get($args->{'radius_request'}->{'User-Name'})) {
+                my $result = $mfa->check_user($args->{'radius_request'}->{'User-Name'}, $$password, $infos->{'device'});
+                if ($result != $TRUE) {
+                    return [ $RADIUS::RLM_MODULE_FAIL, ('Reply-Message' => "MFA verification failed") ];
+                } else {
+                    if ($caller eq "pf::radius::vpn") {
+                        return $args->{'switch'}->returnAuthorizeVPN($args);
+		} else {
+                        return $TRUE;
+                    }
                 }
-                if (exists $pf::config::ConfigAdminRoles{$value}->{'ACTIONS'}->{'SWITCH_LOGIN_READ'}) {
-                    return $switch->returnAuthorizeRead($args);
+            }
+            my @otp = split($mfa->split_char,$$password);
+            $$password = $otp[0];
+            $$otp = $otp[1];
+        } elsif ($mfa->radius_mfa_method eq 'second-password') {
+            if (my $authenticated = $chi->get($args->{'radius_request'}->{'User-Name'}." authenticated")) {
+                if ($authenticated) {
+                    my $result = $mfa->check_user($args->{'radius_request'}->{'User-Name'}, $$password);
+                    if ($result != $TRUE) {
+                        return [ $RADIUS::RLM_MODULE_FAIL, ('Reply-Message' => "MFA verification failed")];
+                    } else {
+                        if ($caller eq "pf::radius::vpn") {
+                            return $args->{'switch'}->returnAuthorizeVPN($args);
+		    } else {
+                            return $TRUE;
+                        }
+                    }
+                } else {
+                    my $device = $chi->get($args->{'radius_request'}->{'User-Name'});
+                    my $result = $mfa->check_user($args->{'radius_request'}->{'User-Name'}, $$password, $device);
+                    if ($result != $TRUE) {
+                        return [ $RADIUS::RLM_MODULE_FAIL, ('Reply-Message' => "MFA verification failed") ];
+                    } else {
+                        if ($caller eq "pf::radius::vpn") {
+                            return $args->{'switch'}->returnAuthorizeVPN($args);
+		    } else {
+                            return $TRUE;
+                        }
+                    }
```

Comment:
There seems to be invalid indentation on many lines there which makes the code very hard to read